core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -87,11 +105,44 @@ public static CloseableIterable<CombinedScanTask> planTasks(
                     + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
                 (1 + file.deletes().size()) * openFileCost);
 
-    return CloseableIterable.transform(
-        CloseableIterable.combine(
-            new BinPacking.PackingIterable<>(splitFiles, splitSize, lookback, weightFunc, true),
-            splitFiles),
-        BaseCombinedScanTask::new);
+    if (preservedPartitionIds != null && !preservedPartitionIds.isEmpty()) {
+      Preconditions.checkArgument(
+          spec != null, "spec can't be null when " + "preservedPartitionIds is not null");
+      StructProjection projectedStruct =
+          StructProjection.create(spec.partitionType(), preservedPartitionIds);
+      Types.StructType projectedPartitionType = projectedStruct.type();
+      ListMultimap<StructLikeWrapper, FileScanTask> groupedFiles =
+          Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
+
+      splitFiles.forEach(
+          f -> {
+            StructLikeWrapper wrapper =
+                StructLikeWrapper.forType(projectedPartitionType)
+                    .set(projectedStruct.copy().wrap(f.file().partition()));
+            groupedFiles.put(wrapper, f);
+          });
+
+      List<Iterable<BaseCombinedScanTask>> groupedTasks =
+          groupedFiles.asMap().values().stream()
+              .map(
+                  t ->
+                      Iterables.transform(
+                          new BinPacking.PackingIterable<>(
+                              CloseableIterable.withNoopClose(t),
+                              splitSize,
+                              lookback,
+                              weightFunc,
+                              true),
+                          BaseCombinedScanTask::new))
+              .collect(Collectors.toList());
+      return CloseableIterable.combine(Iterables.concat(groupedTasks), splitFiles);
+    } else {
+      return CloseableIterable.transform(
+          CloseableIterable.combine(
+              new BinPacking.PackingIterable<>(splitFiles, splitSize, lookback, weightFunc, true),
+              splitFiles),
+          BaseCombinedScanTask::new);
+    }
   }
 
   @SuppressWarnings("unchecked")
```

Comment:
We will to cover `planTaskGroups` as well. It is a new version of the API that we plan to switch to. It is already being used by incremental scans.